File: Problems/784_Letter_Case_Permutation.py
from typing import List

class Solution:
    def letterCasePermutation(self, s: str) -> List[str]:
        res = []
        
        def helper(curr_str: str, curr_index: int):
            if len(curr_str)==len(s):
                res.append(curr_str)
                return
            curr_index += 1               
            if s[curr_index].isalpha():
                helper(curr_str+s[curr_index].lower(), curr_index)
                helper(curr_str+s[curr_index].upper(), curr_index)
            else:
                helper(curr_str + s[curr_index], curr_index)
        helper("", -1)
        return res

if __name__ == '__main__':
    sol = Solution()
    s = "a1b2"
    print(sol.letterCasePermutation(s))

File: Problems/784_Letter_Case_Permutation.cpp
#include<string>
#include<vector>
using namespace std;

class Solution {
public:
    vector<string> res;
    int LEN;
    string S;
    vector<string> letterCasePermutation(string s) {
        LEN = s.length();
        S = s;
        helper("", -1);
        return res;
    }
    void helper(string curr_string, int curr_index){
        if(curr_index == LEN-1){
            res.push_back(curr_string);
            return;
        }
        curr_index += 1;
        if(isalpha(S[curr_index])){
            string temp(curr_string);
            temp+=tolower(S[curr_index]);
            helper(temp, curr_index);
            temp = curr_string;
            temp += toupper(S[curr_index]);
            helper(temp, curr_index);
        }else{
            helper(curr_string+S[curr_index], curr_index);
        }
    }
};